exercise1/terraform/main.tf

```diff
@@ -0,0 +1,537 @@
+  }
+
+}
+
+# Custom Script Extension to install NGINX and configure it
+resource "azurerm_virtual_machine_extension" "ex1_vm_extension" {
+  name                       = "nginx_setup"
+  virtual_machine_id         = azurerm_virtual_machine.ex1_vm.id
+  publisher                  = "Microsoft.Azure.Extensions"
+  type                       = "CustomScript"
+  type_handler_version       = "2.0"
+  automatic_upgrade_enabled  = false
+  auto_upgrade_minor_version = false
+
+  settings = <<SETTINGS
+    {
+        "script": "${base64encode(file(var.nginxConfig))}"
+    }
+SETTINGS
+
+  depends_on = [azurerm_virtual_machine.ex1_vm]
+}
+
+#___________________________ App Gateway network resources ______________________________
+resource "azurerm_subnet" "ex1_subnet_app_gw" {
+  name                 = "${var.rg_name}_subnet_app_gw"
+  resource_group_name  = azurerm_resource_group.ex1.name
+  virtual_network_name = azurerm_virtual_network.ex1_vnet.name
+  address_prefixes     = ["10.0.3.0/24"]
+}
+
+resource "azurerm_network_security_group" "ex1_app_gw_netsecg" {
+  name                = "${var.rg_name}_app_gw"
+  location            = azurerm_resource_group.ex1.location
+  resource_group_name = azurerm_resource_group.ex1.name
+}
+
+resource "azurerm_subnet_network_security_group_association" "ex1_secg_asso_app_gw" {
+  subnet_id                 = azurerm_subnet.ex1_subnet_app_gw.id
+  network_security_group_id = azurerm_network_security_group.ex1_app_gw_netsecg.id
+}
+
+resource "azurerm_network_security_rule" "https_rule_app_gw" {
+  name                        = "AllowHTTPS"
+  priority                    = 1002
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "443"
+  source_address_prefix       = "*"
+  destination_address_prefix  = "*"
+  resource_group_name         = azurerm_resource_group.ex1.name
+  network_security_group_name = azurerm_network_security_group.ex1_app_gw_netsecg.name
+}
+# required ports by application gateway please see https://learn.microsoft.com/en-us/azure/application-gateway/configuration-infrastructure#:~:text=V2%3A%20Ports%2065200%2D65535
+resource "azurerm_network_security_rule" "health_probe_inbound" {
+  name                        = "AllowHealthProbe"
+  priority                    = 1003
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "65200-65535"
+  source_address_prefix       = "*"
+  destination_address_prefix  = "*"
+  resource_group_name         = azurerm_resource_group.ex1.name
+  network_security_group_name = azurerm_network_security_group.ex1_app_gw_netsecg.name
+}
+
+# Deny all inbound traffic not explicitly allowed
+resource "azurerm_network_security_rule" "deny_all_inbound_app_gw" {
```

Comment:
You need this allow before this deny:
AzureLoadBalancer	Any	Any	Any	Any	Allow

Reply:
added `lb_inbound` if i understood correctly